File: lib/Service/Clam.php
<?php

declare(strict_types=1);

// SPDX-FileCopyrightText: 2026 Cristian Casapu
// SPDX-License-Identifier: AGPL-3.0-or-later

namespace OCA\Sentinel\Service;

use Psr\Log\LoggerInterface;

/**
 * Asking ClamAV about a specific file, at a specific moment.
 *
 * Not a virus scanner for the server — Nextcloud has files_antivirus for that,
 * and scanning every upload is a different job with a different cost. This is
 * the narrow thing Sentinel needs: when something has already looked wrong for
 * another reason, put the handful of files that caused the suspicion in front
 * of a scanner and see whether it recognises them.
 *
 * It talks to clamd directly over its socket. Shelling out to clamscan would
 * load the entire signature database for every file, which takes a quarter of a
 * minute and a gigabyte of memory; clamd already has it loaded.
 */
class Clam {
	private const CHUNK = 65536;

	public function __construct(
		private Settings $settings,
		private LoggerInterface $logger,
	) {
	}

	/** Is there anything at the other end of the socket? */
	public function available(): bool {
		if (!$this->settings->clamEnabled()) {
			return false;
		}
		return $this->ping() !== null;
	}

	/**
	 * What clamd says about itself: version and how old its signatures are.
	 *
	 * A scanner running on last year's signatures is worse than none, because
	 * it answers "clean" with the same confidence either way.
	 *
	 * @return array<string, mixed>
	 */
	public function state(): array {
		if (!$this->settings->clamEnabled()) {
			return ['enabled' => false, 'reachable' => false, 'reason' => 'switched off'];
		}

		$socket = $this->settings->clamSocket();
		if (!file_exists($socket)) {
			return [
				'enabled' => true,
				'reachable' => false,
				'socket' => $socket,
				'reason' => 'no socket at that path',
			];
		}

		$version = $this->ping();
		if ($version === null) {
			return ['enabled' => true, 'reachable' => false, 'socket' => $socket, 'reason' => 'nothing answered'];
		}

		// "ClamAV 1.0.7/27321/Mon Sep 8 09:12:03 2026"
		$parts = explode('/', $version);
		$signatureDate = isset($parts[2]) ? strtotime(trim($parts[2])) : false;

		return [
			'enabled' => true,
			'reachable' => true,
			'socket' => $socket,
			'version' => trim($parts[0] ?? $version),
			'signatures' => isset($parts[1]) ? (int)$parts[1] : 0,
			'signaturesAt' => $signatureDate === false ? 0 : $signatureDate,
			'signatureAgeDays' => $signatureDate === false ? null : (int)floor((time() - $signatureDate) / 86400),
		];
	}

	/**
	 * Scan one stream.
	 *
	 * @param resource $stream
	 * @return array{scanned: bool, infected: bool, name: string}
	 */
	public function scanStream($stream, int $maxBytes): array {
		$socket = $this->connect();
		if ($socket === null) {
			return ['scanned' => false, 'infected' => false, 'name' => ''];
		}

		try {
			fwrite($socket, "zINSTREAM\0");
			$sent = 0;
			while (!feof($stream) && $sent < $maxBytes) {
				$chunk = fread($stream, min(self::CHUNK, $maxBytes - $sent));
				if ($chunk === false || $chunk === '') {
					break;
				}
				$sent += strlen($chunk);
				fwrite($socket, pack('N', strlen($chunk)) . $chunk);
			}
			// A zero-length chunk is how INSTREAM says "that is all of it".
			fwrite($socket, pack('N', 0));

			$answer = trim((string)stream_get_contents($socket));
			return $this->readAnswer($answer);
		} catch (\Throwable $e) {
			$this->logger->warning('Sentinel could not scan a stream with ClamAV', ['exception' => $e]);
			return ['scanned' => false, 'infected' => false, 'name' => ''];
		} finally {
			fclose($socket);
		}
	}

	/**
	 * Scan a file on the local disk.
	 *
	 * @return array{scanned: bool, infected: bool, name: string}
	 */
	public function scanPath(string $path): array {
		if (!is_readable($path)) {
			return ['scanned' => false, 'infected' => false, 'name' => ''];
		}
		$handle = @fopen($path, 'rb');
		if (!is_resource($handle)) {
			return ['scanned' => false, 'infected' => false, 'name' => ''];
		}
		try {
			return $this->scanStream($handle, $this->settings->clamMaxBytes());
		} finally {
			fclose($handle);
		}
	}

	/** @return array{scanned: bool, infected: bool, name: string} */
	private function readAnswer(string $answer): array {
		// "stream: OK" or "stream: Eicar-Test-Signature FOUND" or "... ERROR"
		$answer = rtrim($answer, "\0");
		if ($answer === '' || str_ends_with($answer, 'ERROR')) {
			return ['scanned' => false, 'infected' => false, 'name' => $answer];
		}
		if (str_ends_with($answer, 'FOUND')) {
			$name = trim(substr($answer, strpos($answer, ':') + 1));
			$name = trim(substr($name, 0, -strlen('FOUND')));
			return ['scanned' => true, 'infected' => true, 'name' => $name];
		}
		return ['scanned' => true, 'infected' => false, 'name' => ''];
	}

	private function ping(): ?string {
		$socket = $this->connect();
		if ($socket === null) {
			return null;
		}
		try {
			fwrite($socket, "zVERSION\0");
			$answer = trim((string)stream_get_contents($socket));
			return $answer === '' ? null : rtrim($answer, "\0");
		} catch (\Throwable) {
			return null;
		} finally {
			fclose($socket);
		}
	}

	/** @return resource|null */
	private function connect() {
		$path = $this->settings->clamSocket();
		if ($path === '') {
			return null;
		}

		$target = str_contains($path, ':') && !str_starts_with($path, '/')
			? 'tcp://' . $path
			: 'unix://' . $path;

		$socket = @stream_socket_client($target, $code, $error, 5);
		if (!is_resource($socket)) {
			return null;
		}
		stream_set_timeout($socket, 30);
		return $socket;
	}
}

File: edr/sentinel_edr/clam.py
# SPDX-FileCopyrightText: 2026 Cristian Casapu
# SPDX-License-Identifier: AGPL-3.0-or-later
"""Asking clamd about one file, at the moment it becomes interesting.

Not a scanner for everything that arrives — that is a different job with a
different cost, and Nextcloud has an app for it. This is the narrow use: once
something already looks wrong for other reasons, put the handful of files that
caused the suspicion, and the program that wrote them, in front of a scanner.

Spoken to over its socket rather than through clamscan, which loads the whole
signature database for every single file it is asked about.
"""

from __future__ import annotations

import socket
import struct
from dataclasses import dataclass

CHUNK = 65536


@dataclass(frozen=True)
class Verdict:
    scanned: bool
    infected: bool
    name: str = ""

    @property
    def useful(self) -> bool:
        return self.scanned


UNKNOWN = Verdict(scanned=False, infected=False)


class Clam:
    def __init__(self, address: str, timeout: float = 30.0, max_bytes: int = 32 * 1024 * 1024) -> None:
        self.address = address
        self.timeout = timeout
        self.max_bytes = max_bytes

    def _connect(self) -> socket.socket | None:
        try:
            if self.address.startswith("/"):
                sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
                sock.settimeout(self.timeout)
                sock.connect(self.address)
            else:
                host, _, port = self.address.rpartition(":")
                sock = socket.create_connection((host or "127.0.0.1", int(port or 3310)), self.timeout)
                sock.settimeout(self.timeout)
            return sock
        except OSError:
            return None

    def version(self) -> str | None:
        sock = self._connect()
        if sock is None:
            return None
        try:
            sock.sendall(b"zVERSION\0")
            answer = sock.recv(512).decode("utf-8", "replace").strip().rstrip("\0")
            return answer or None
        except OSError:
            return None
        finally:
            sock.close()

    def scan_fd(self, fd: int) -> Verdict:
        """Scan through an open descriptor.

        The descriptor, not the path: between noticing a file and reading it,
        the name may already belong to something else, and it is precisely the
        interesting cases where that is likely.
        """
        import os

        sock = self._connect()
        if sock is None:
            return UNKNOWN
        try:
            sock.sendall(b"zINSTREAM\0")
            sent = 0
            while sent < self.max_bytes:
                chunk = os.pread(fd, min(CHUNK, self.max_bytes - sent), sent)
                if not chunk:
                    break
                sent += len(chunk)
                sock.sendall(struct.pack("!I", len(chunk)) + chunk)
            sock.sendall(struct.pack("!I", 0))
            return self._read(sock)
        except OSError:
            return UNKNOWN
        finally:
            sock.close()

    def scan_path(self, path: str) -> Verdict:
        import os

        try:
            fd = os.open(path, os.O_RDONLY | os.O_NOFOLLOW)
        except OSError:
            return UNKNOWN
        try:
            return self.scan_fd(fd)
        finally:
            os.close(fd)

    def _read(self, sock: socket.socket) -> Verdict:
        try:
            answer = sock.recv(4096).decode("utf-8", "replace").strip().rstrip("\0")
        except OSError:
            return UNKNOWN
        if not answer or answer.endswith("ERROR"):
            return UNKNOWN
        if answer.endswith("FOUND"):
            name = answer.split(":", 1)[-1].strip()
            name = name[: -len("FOUND")].strip()
            return Verdict(scanned=True, infected=True, name=name)
        return Verdict(scanned=True, infected=False)
